### PR TITLE
feat(indy-vdr): add indy-vdr package and indy vdr pool

### PR DESCRIPTION
**Pull request by Vickysomtee**
This PR implements the new IndyVdr Pool Package

**Work funded by the government of Ontario**

### CODE REVIEW

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:70]:
```suggestion
   * Get the most appropriate pool for the given did. If the did is a qualified indy did, the pool will be determined based on the namespace. If it is a legacy unqualified indy did, the pool will be determined based on the algorithm as described in this document:
   * https://docs.google.com/document/d/109C_eMsuZnTnYe2OAd02jAts1vC4axwEKIq7_4dnNVA/edit
   */
```

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:74]: The namespace can contain multiple colons. 

E.g. the following dids are all correct:

- `did:indy:sovrin:builder:xxx` -> namespace = `sovrin:builder`
- `did:indy:bcovrin:xxx` -> namespace = `bcovrin`.

We should proabbly use a regex to match the did. If you store the followign const in a separate file:

```
const DID_INDY_REGEX = /^did:indy:((?:[a-z][_a-z0-9-]*)(?::[a-z][_a-z0-9-]*)):([1-9A-HJ-NP-Za-km-z]{21,22})$/
```

Then you can use it like this:

```ts
const match = did.match(DID_INDY_REGEX)

// If we have a match it means the did starts with did:indy and is a qualified indy did
if (match) {
  const [, namespace] = match
} else {
  // legacy way
}
```

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:80]: Can we also update this to the indy vdr error?

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:149]: I think we can update this to just return the pool

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:104]: Same here

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:78]: I think we can return the pool on the top level. So just return pool instead of return { pool }

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:182]: We can use `pool.indyNamespace` directly instead of using `pool.config` I think

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [reply packages/indy-vdr/src/pool/IndyVdrPoolService.ts:74]: We can then also just use `this.getPoolForNamespace` and reuse that logic

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:221]: This is from how the indy-sdk worked and it would trhow a ledger not found error. Now, it will return the json and the `data` filed will be `null` (you already do that check).

So this can be simlified to:

```suggestion
 throw error
```

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:200]: We should probably also create an IndyVdrNotFoundError error based on the indy vdr error and throw that here. Aslo can you make the error message a bit more descriptive?

**Review by TimoGlastra — CHANGES_REQUESTED** [review]
Great work @Vickysomtee! Left some comments mostly on things that can be done a bit differently in the indy vdr library compared to the indy sdk

File: packages/core/src/index.ts
Comment (TimoGlastra) [inline packages/core/src/index.ts:30]: The cache directory doesn't exist anymore in `main`. It's now in `packages/core/src/module/cache`. However the needed files should be exported already

File: packages/indy-vdr/README.md
Comment (TimoGlastra) [inline packages/indy-vdr/README.md:9]: The readme is still for the action menu pacakges. Can you update all references?

File: packages/indy-vdr/package.json
Comment (TimoGlastra) [inline packages/indy-vdr/package.json:5]: Can you align the version with version of the core package?

File: packages/indy-vdr/package.json
Comment (TimoGlastra) [inline packages/indy-vdr/package.json:33]: Can you align all `@aries-framework/*` versions with the core one?

File: packages/indy-vdr/src/error/IndyVdrNotConfiguredError.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/error/IndyVdrNotConfiguredError.ts:3]:
```suggestion
export class IndyVdrNotConfiguredError extends IndyVdrError {
```

File: packages/indy-vdr/src/pool/DidIdentifier.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/DidIdentifier.ts:1]: can you rename this file to `didIdentifier` instead of `DidIdentifier`? We use Pascal case for files that contains classes only. 

See here how to commit a case only name change in git (it's a pain): https://stackoverflow.com/questions/17683458/how-do-i-commit-case-sensitive-only-filename-changes-in-git

File: packages/indy-vdr/src/pool/IndyVdrPool.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPool.ts:41]:
```suggestion
```

File: packages/indy-vdr/src/pool/IndyVdrPool.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPool.ts:73]: This should use IndyVdrError

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:16]: These are not used anymore

File: packages/indy-vdr/README.md
Comment (genaris) [inline packages/indy-vdr/README.md:9]: This README will need to be updated with info about indy-vdr

File: packages/indy-vdr/src/index.ts
Comment (genaris) [inline packages/indy-vdr/src/index.ts:2]: How will this work when we want to use this package in React Native? Should we add the require('indy-vdr-test-react-native') to the catch block?

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:177]: This can be removed. The nameSpace is already required

File: packages/indy-vdr/src/pool/IndyVdrPoolService.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPoolService.ts:182]: Should be `IndyVdrError` (not found is for when it's not found on the ledger)

File: packages/indy-vdr/src/pool/IndyVdrPool.ts
Comment (genaris) [inline packages/indy-vdr/src/pool/IndyVdrPool.ts:72]: I guess you can use IndyVdrError now it is created

File: packages/indy-vdr/tests/indy-vdr-pool.e2e.test.ts
Comment (TimoGlastra) [inline packages/indy-vdr/tests/indy-vdr-pool.e2e.test.ts:241]: This test doesn't test what it says it does. I think we can just remove it for now

**Review by genaris — DISMISSED** [review]
Looks good! I've left a few comments and questions about the initialization.

Probably as part of another PR (when adding consumers to Indy VDR Pool like the resolvers and registry), but should we create it as a module that registers services and also initializes the binding library within its `register` method?

File: packages/indy-vdr/src/index.ts
Comment (TimoGlastra) [reply packages/indy-vdr/src/index.ts:2]: I was thinking we could either do it with the catch as you say, or we detect the platform. Or would you prefer the manual approach? (so you provide it to the module). That's also fine for me.

File: packages/indy-vdr/src/pool/IndyVdrPool.ts
Comment (TimoGlastra) [reply packages/indy-vdr/src/pool/IndyVdrPool.ts:85]: Yes. it hangs indefinitely. should create an issue for it in the indy-vdr repo. @blu3beri do you still know the details?

Comment (blu3beri) [reply packages/indy-vdr/src/pool/IndyVdrPool.ts:85]: Ah yeah, apparently, almost a year ago now, I create an issue for this.
https://github.com/hyperledger/indy-vdr/issues/92

Comment (blu3beri) [reply packages/indy-vdr/src/pool/IndyVdrPool.ts:85]: I can investigate this further if it's required. The reasoning (no write-access to lock) might be completely incorrect.

File: packages/indy-vdr/package.json
Comment (genaris) [inline packages/indy-vdr/package.json:41]: You will also need to update this line to "typescript": "~4.9.4" to align with other packages. Also for the rimraf (you'll need to remove the -rf parameter in the clean test script)

File: packages/indy-vdr/tests/indy-vdr-pool.e2e.test.ts
Comment (karimStekelenburg) [inline packages/indy-vdr/tests/indy-vdr-pool.e2e.test.ts:123]: @TimoGlastra this was originally added by me. I assumed we could remove this, as the CI uses a clean ledger for every run. Is my assumption correct?

**Review by karimStekelenburg — APPROVED** [review]
Looks like these are mostly style changes and they look good to me. I'm assuming the README will be done in a follow-up PR, so I'll approve. If not, this needs to be added before merging.

In order to do so, just have a look at the READMEs of other packages inside this repository. They'll give you an idea of what is required.

File: packages/indy-vdr/tests/indy-vdr-pool.e2e.test.ts
Comment (TimoGlastra) [reply packages/indy-vdr/tests/indy-vdr-pool.e2e.test.ts:123]: Yeah we can. But then subsequent test runs (local) will fail I think? So I thought there's not harm in keeping it in.

File: packages/indy-vdr/tests/setup.ts
Comment (TimoGlastra) [inline packages/indy-vdr/tests/setup.ts:1]: This is not needed, as we don't use `class-validator` or `class-transformer`

```suggestion
```

File: packages/indy-vdr/src/pool/didIdentifier.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/didIdentifier.ts:1]: Maybe we can move this to the `utlils/did.ts` file next to the `ABBREVIATED_VERKEY_REGEX`?

File: packages/indy-vdr/package.json
Comment (TimoGlastra) [inline packages/indy-vdr/package.json:41]:
```suggestion
    "rimraf": "^4.0.7",
```

File: packages/indy-vdr/package.json
Comment (TimoGlastra) [inline packages/indy-vdr/package.json:40]:
```suggestion
```

File: packages/indy-vdr/package.json
Comment (TimoGlastra) [inline packages/indy-vdr/package.json:38]:
```suggestion
```

File: packages/indy-vdr/package.json
Comment (TimoGlastra) [inline packages/indy-vdr/package.json:35]:
```suggestion
```

File: packages/indy-vdr/package.json
Comment (TimoGlastra) [inline packages/indy-vdr/package.json:33]:
```suggestion
```

File: .eslintrc.js
Comment (TimoGlastra) [inline .eslintrc.js:25]: We shouldn't need this I think. What is the error you're getting?



```suggestion
```

Comment (Vickysomtee) [reply .eslintrc.js:25]: I am getting this no-extranous-dependencies error from lint

I figured out it's from this

```
try {
  require('indy-vdr-test-nodejs')
} catch (error) {
  throw new Error('Error registering nodejs bindings for Indy VDR')
}

```

Comment (TimoGlastra) [reply .eslintrc.js:25]: ah you can add an eslint ignore above the import, it's fine in this case

Comment (Vickysomtee) [reply .eslintrc.js:25]: This is the full error message

`'indy-vdr-test-nodejs' should be listed in the project's dependencies, not devDependencies  import/no-extraneous-dependencies`

Comment (TimoGlastra) [reply .eslintrc.js:25]: This should work:

```
try {
  // eslint-disable-next-line import/no-extraneous-dependencies
  require('indy-vdr-test-nodejs')
} catch (error) {
  throw new Error('Error registering nodejs bindings for Indy VDR')
}
```

**Review by TimoGlastra — APPROVED** [review]

File: packages/indy-vdr/src/pool/IndyVdrPool.ts
Comment (TimoGlastra) [inline packages/indy-vdr/src/pool/IndyVdrPool.ts:174]: Cast to unknown first until we have a new release  for indy-vdr

```suggestion
     // FIME: remove cast when https://github.com/hyperledger/indy-vdr/pull/142 is released
    const acceptanceMechanisms = acceptanceMechanismResponse.result.data as unknown as AcceptanceMechanisms
```

File: packages/indy-vdr/package.json
Comment (TimoGlastra) [inline packages/indy-vdr/package.json:5]:
```suggestion
  "version": "0.3.3",
  "private": true,
```

**Comment by genaris** [comment]
I think it would be nice to use @hyperledger/indy-vdr-* packages now that they are going to be released in NPM after the v0.4.0-dev.2 release (https://github.com/hyperledger/indy-vdr/releases/tag/v0.4.0-dev.2)

**Comment by TimoGlastra** [comment]
> I think it would be nice to use @hyperledger/indy-vdr-* packages now that they are going to be released in NPM after the v0.4.0-dev.2 release ([hyperledger/indy-vdr@`v0.4.0-dev.2` (release)](https://github.com/hyperledger/indy-vdr/releases/tag/v0.4.0-dev.2))

It doesn't quite work yet, so I'm going to merge this once CI passes and once the build is working we can make a follow up PR.

**Pull request merged**
